File: architecture/common/src/main/java/com/pacific/common/rx/CompletableUtil.java
package com.pacific.common.rx;

import io.reactivex.CompletableTransformer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import java.util.concurrent.Executor;

public class CompletableUtil {

  private CompletableUtil() {
    throw new UnsupportedOperationException();
  }

  public static CompletableTransformer io() {
    return completable -> completable.subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread());
  }

  public static CompletableTransformer computation() {
    return completable -> completable.subscribeOn(Schedulers.computation())
        .observeOn(AndroidSchedulers.mainThread());
  }

  public static CompletableTransformer trampoline() {
    return completable -> completable.subscribeOn(Schedulers.trampoline())
        .observeOn(AndroidSchedulers.mainThread());
  }

  public static CompletableTransformer newThread() {
    return completable -> completable.subscribeOn(Schedulers.newThread())
        .observeOn(AndroidSchedulers.mainThread());
  }

  public static CompletableTransformer single() {
    return completable -> completable.subscribeOn(Schedulers.single())
        .observeOn(AndroidSchedulers.mainThread());
  }

  public static CompletableTransformer from(final Executor executor) {
    return completable -> completable.subscribeOn(Schedulers.from(executor))
        .observeOn(AndroidSchedulers.mainThread());
  }
}

File: architecture/arch-kotlin/src/main/java/com/pacific/arch/kotlin/rx/CompletableUtil.kt
package com.pacific.arch.kotlin.rx

import io.reactivex.CompletableTransformer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.Executor

class CompletableUtil private constructor() {
    companion object {

        @JvmStatic
        fun io() = CompletableTransformer {
            it.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        }

        @JvmStatic
        fun computation() = CompletableTransformer {
            it.subscribeOn(Schedulers.computation()).observeOn(AndroidSchedulers.mainThread())
        }

        @JvmStatic
        fun trampoline() = CompletableTransformer {
            it.subscribeOn(Schedulers.trampoline()).observeOn(AndroidSchedulers.mainThread())
        }

        @JvmStatic
        fun newThread() = CompletableTransformer {
            it.subscribeOn(Schedulers.newThread()).observeOn(AndroidSchedulers.mainThread())
        }

        @JvmStatic
        fun single() = CompletableTransformer {
            it.subscribeOn(Schedulers.single()).observeOn(AndroidSchedulers.mainThread())
        }

        @JvmStatic
        fun from(executor: Executor) = CompletableTransformer {
            it.subscribeOn(Schedulers.from(executor)).observeOn(AndroidSchedulers.mainThread())
        }
    }
}
